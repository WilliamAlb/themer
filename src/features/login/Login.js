import React,{useState,useEffect} from "react";
import { useSelector } from "react-redux";
import styled from "styled-components";
import { getAuth, signInWithEmailAndPassword,createUserWithEmailAndPassword } from "@firebase/auth";
import { useAuthState } from 'react-firebase-hooks/auth';

import { collection,setDoc,doc } from "firebase/firestore/lite";

import { selectTheme } from "../slices/currentThemeSlice";
import './loginStyle.css';
import * as S from './login.style';
import { auth } from "../../util/firebase";
import { Route, useNavigate } from "react-router";
import { database } from "../../util/firebase";
import { NavLink } from "react-router-dom";
import { routes } from "../../util/routes";

export function Login(){

    
    const [username,setUsername] = useState('');
    const [password,setPassword] = useState('');

    const handleSubmit =(e)=>{
        e.preventDefault();
        signInWithEmailAndPassword(auth,username,password)
        .then(
            (userCredentials)=>{
                setUsername('');
                setPassword('');
            }).catch(
                (error)=> console.log(error)
                );

    }
    
    const registerWithEmailAndPassword = async (e) => {
        e.preventDefault();
        try {
          const res = await createUserWithEmailAndPassword(auth,username, password);
          const user = res.user;
          const docData = doc(database,'users','one');
          
          await setDoc(docData,{
            uid: user.uid,
            displayName:'fuckface',
            email:username,
          });
        } catch (err) {
          console.error(err);
          console.log(err.message);
        }
      };

    return (
        <S.LoginSection>
            <div className='loginContent'>
                <S.H1>Login to Themer</S.H1>
                <form className='loginForm' onSubmit={handleSubmit}>
                    <label htmlFor='usernameInput'>Email</label>
                    <br/>
                    <S.Input id='usernameInput' type='text' value={username} onChange={({target})=>setUsername(target.value)}></S.Input>
                    <br/>
                    <label htmlFor='passwordInput'>Password</label>
                    <br/>
                    <S.Input id='passwordInput' type='password'value={password} onChange={({target})=>setPassword(target.value)}></S.Input>
                    <br/>
                    <S.SubmitButton type='submit' value='Login'></S.SubmitButton>
                </form>
                <p>Don't have an account? {<NavLink to={routes.register}>Register</NavLink>} </p>
            </div>
            <div className='loginSplash'></div>
        </S.LoginSection>
    );
}