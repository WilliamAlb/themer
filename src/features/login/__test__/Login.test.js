import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { render, screen, waitForDomChange } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import * as FirBaseAuthModule from "@firebase/auth";
import { BrowserRouter as Router } from "react-router-dom";
import { Login } from "../Login";

const loginMockFn = jest.fn();
const mockServer = setupServer(rest.post('https://identitytoolkit.googleapis.com/v1/*',(req,res,context) => {
  loginMockFn(req.body)
  return res(context.json({username: "shit"}))
}))

beforeAll(() => {
  mockServer.listen()
})

afterEach(() => {
  mockServer.resetHandlers()
})

afterAll(() => {
  mockServer.close()
})

const setup = () => {
  const component = (
    <Router>
      <Login />
    </Router>
  );
  render(component);
};

const getEmailInput = () => screen.getByLabelText(/email/i);
const getPasswordInput = () => screen.getByLabelText(/password/i);
const getSubmitButton = () => screen.getByDisplayValue(/login/i);

describe("login", () => {
  it("when empty state button should be disabled", () => {
    setup();
    // expect(getSubmitButton()).toBeDisabled();
  });

  it("On goto register page link click should navigate to register", () => {});

  it("On firebase error should display popup", async () => {
    setup();

    userEvent.type(getEmailInput(), "email@gmail.com");
    userEvent.type(getPasswordInput(), "Aa12345!");
    expect(getSubmitButton()).not.toBeDisabled();

    userEvent.click(getSubmitButton());

    // IMPLEMENT ON ERROR INSTEAD OF SUCCESS
    await waitForDomChange(() => {
      expect(getEmailInput()).toHaveValue('')
      expect(getPasswordInput()).toHaveValue('')
    });

    expect(loginMockFn.mock.calls[0]).toMatchSnapshot()
  });

  it("When form is valid end fire submit should call firebase", async () => {
    const mockFn = jest
      .spyOn(FirBaseAuthModule, "signInWithEmailAndPassword")
      .mockImplementation(() => {
        return new Promise((res, rej) => res());
      });
    setup();

    userEvent.type(getEmailInput(), "email@gmail.com");
    userEvent.type(getPasswordInput(), "Aa12345!");
    expect(getSubmitButton()).not.toBeDisabled();

    userEvent.click(getSubmitButton());
    const [auth, email, password] = mockFn.mock.calls[0];
    expect({ auth, email, password }).toMatchSnapshot();
    expect(password).toMatchInlineSnapshot(`"Aa12345!"`);

    await waitForDomChange(() => {
      expect(getEmailInput()).toHaveValue('')
      expect(getPasswordInput()).toHaveValue('')
    })

  });

  
});
