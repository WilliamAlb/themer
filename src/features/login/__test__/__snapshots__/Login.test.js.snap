// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login On firebase error should display popup 1`] = `
Array [
  Object {
    "email": "email@gmail.com",
    "password": "Aa12345!",
    "returnSecureToken": true,
  },
]
`;

exports[`login When form is valid end fire submit should call firebase 1`] = `
Object {
  "auth": Object {
    "apiKey": "AIzaSyBAkNPiG1LX2-PlScZm_TOpTUGk6t3IBL4",
    "appName": "[DEFAULT]",
    "authDomain": "themer-9acfa.firebaseapp.com",
    "currentUser": undefined,
  },
  "email": "email@gmail.com",
  "password": "Aa12345!",
}
`;
